### PR TITLE
feat(garbage-collector): batch parquet existence checks to catalog

### LINKED ISSUE #7784: GC should batch get_by_object_store_id() lookups
In a cluster where we're clearing a backlog of orphaned files in object storage (via the garbage collector) it's causing a relatively high query workload against the catalog:

plot

The `parquet.get_by_object_store_id()` calls are only made by GC, and are responsible for **~70% of the catalog load** for the last few days in a cluster processing several thousand writes/s:

https://github.com/influxdata/influxdb_iox/blob/36c1d23e19d3dd258bdc7dd350082355bbc3ed82/garbage_collector/src/objectstore/checker.rs#L92

Instead of performing one query per ID, we should batch the ID lookups (for example, look up 500 IDs at once). This will significantly reduce the catalog load!

### PR DESCRIPTION
**Pull request by philjb**
---

* feat(garbage-collector): batch parquet existence checks to catalog
    
    The core feature of this PR is batching the existence checks of parquet
    files in object store against the catalog. Before, there was 1 catalog
    query per each parquet file in object store. This can be a lot of
    requests.
    
    This PR can perform one query of at most 100 parquet file uuids against
    the catalog in one query. A hundred seems like a decent starting place.
    
    The batch may not reach 100 because there is also a timeout on receiving
    object store meta objects from the object store lister thread. That
    timeout is set to 100 milliseconds. If more than 100 are received, they
    are batched into 100 for the catalog.
    
    Additionally, this PR includes surrounding code changes to make it more
    idiomatic (but not perfect). It follows up some suggested work from
     #7652 for watching for shutdown on the threads.
    
* fixes #7784

### CODE REVIEW

File: compactor/src/components/skipped_compactions_source/mod.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ pub mod mock;
 
 #[async_trait]
 pub trait SkippedCompactionsSource: Debug + Display + Send + Sync {
-    /// Get skipped partition recrod for a given partition
+    /// Get skipped partition record for a given partition
```
Comment (philjb) [inline compactor/src/components/skipped_compactions_source/mod.rs:11]: driveby

File: garbage_collector/src/lib.rs [hunk 2]
```diff
@@ -120,17 +120,31 @@ impl GarbageCollector {
...
+        let cat = Arc::clone(&catalog);
+        let sdt = shutdown.clone();
+        let cutoff = chrono::Duration::from_std(sub_config.objectstore_cutoff).map_err(|e| {
+            Error::CutoffError {
+                message: e.to_string(),
+            }
+        })?;
+
+        let os_checker = tokio::spawn(async move {
+            select! {
```
Comment (philjb) [inline garbage_collector/src/lib.rs:133]: Like what was done over here but not for all the spawned threads yet. (I haven't forgotten!)

https://github.com/influxdata/influxdb_iox/pull/7652/files#r1179156749

File: garbage_collector/src/objectstore/checker.rs [hunk 3]
```diff
@@ -22,16 +23,25 @@ pub enum Error {
...
     #[snafu(display("The deleter task exited unexpectedly"))]
     DeleterExited {
         source: tokio::sync::mpsc::error::SendError<ObjectMeta>,
     },
 }
 
+// The number of parquet files we will ask the catalog to look for at once.
+// todo(pjb): I have no idea what's a good value here to amortize the request. More than 1 is a start.
+const CATALOG_BATCH_SIZE: usize = 100;
+const RECEIVE_TIMEOUT_MS: u64 = 100; // This may not be long enough to collect many objects.
```
Comment (philjb) [inline garbage_collector/src/objectstore/checker.rs:40]: These values are guess...

I don't see any metrics registry in the garbage collector or at least I don't know where to find it to see how these numbers will work out - phase 2 I guess.

File: garbage_collector/src/objectstore/checker.rs [hunk 4]
```diff
@@ -40,96 +50,159 @@ pub(crate) async fn perform(
...
+                        batch.push(item);
                     }
                     None => {
                         // The channel has been closed unexpectedly
                         return Err(Error::ChannelClosed);
                     }
                 }
             }
+            Err(_) => {
+                timedout = true;
```
Comment (philjb) [inline garbage_collector/src/objectstore/checker.rs:74]: Is there another idiom that is clearer for this sort of scenario?

File: garbage_collector/src/objectstore/checker.rs (hunk 3 above)
Comment (Joe-Blount) [reply garbage_collector/src/objectstore/checker.rs:40]: I like these values as a starting place.  Specifically, I like having both criteria.  

To me, this works out to:
// group everything you can for 100ms, because that's not so much of a delay that it would cause issues,
// but if you manage to get a huge number of file ids, stop accumulating at 100.

We may tweak the numbers, but that sentiment feels reasonable for how to think about them.

File: data_types/src/lib.rs [hunk 5]
```diff
@@ -681,6 +681,14 @@ pub struct ParquetFileExists {
     pub exists: i32,
 }
 
+/// Struct to contain row result from a SELECT todo 1 as exists FROM ... query to support
+/// testing the existence of a parquet file.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, sqlx::FromRow)]
```
Comment (crepererum) [inline data_types/src/lib.rs:686]: If you only need this for the PG impl, this should to to `iox_catalog::postgres`.

File: garbage_collector/src/objectstore/checker.rs (hunk 3 above)
Comment (crepererum) [inline garbage_collector/src/objectstore/checker.rs:40]: Instead of `_MS: u64`, I think this should be `: Duration`. `Duration::from_millis` is `const` and can be used to create constants.

File: garbage_collector/src/objectstore/checker.rs [hunk 6]
```diff
@@ -22,16 +23,25 @@ pub enum Error {
...
 }
 
+// The number of parquet files we will ask the catalog to look for at once.
+// todo(pjb): I have no idea what's a good value here to amortize the request. More than 1 is a start.
+const CATALOG_BATCH_SIZE: usize = 100;
+const RECEIVE_TIMEOUT_MS: u64 = 100; // This may not be long enough to collect many objects.
+
 pub(crate) type Result<T, E = Error> = std::result::Result<T, E>;
 
 pub(crate) async fn perform(
```
Comment (crepererum) [inline garbage_collector/src/objectstore/checker.rs:44]: If you don't wanna deal with the `select!` and the actual impl. in one method, you can also pull the actual code into its own method (e.g. `perform_impl`) and ahve the `perform` just call the `select!` macro and the impl method.

File: garbage_collector/src/objectstore/checker.rs (hunk 4 above)
Comment (crepererum) [reply garbage_collector/src/objectstore/checker.rs:74]: Yes:

```rust
let timeout = match maybe_item {
    Ok(res) => {
        ...
        false
    }
    Err(_) => {
        true
    }
};
```

or

```rust
let timeout = maybe_item.is_err();
```

(`is_err` is not consuming)

File: iox_catalog/src/postgres.rs [hunk 7]
```diff
@@ -1491,6 +1491,26 @@ WHERE object_store_id = $1;
         Ok(true)
     }
 
+    async fn exists_by_object_store_id_batch(
```
Comment (crepererum) [inline iox_catalog/src/postgres.rs:1494]: Can we just make the `exists_by_object_store_id` batched and not add yet another method? I think the non-batch method is only used by tests and they can easily use a 1-batch.

Also your new interface is not covered by a test (should be done in `interface.rs`, there are generic tests).

Also I think the method should just return a `ParquetFileID` or something, not a struct w/ a single element (this is just an implementation detail of the SQL wrapper).

File: garbage_collector/src/objectstore/checker.rs [hunk 8]
```diff
@@ -40,96 +50,153 @@ pub(crate) async fn perform(
...
-            "Scheduling file for deletion",
-        );
     }
 
-    Ok(true)
+    // we have a Vec of uuids for the files we _do not_ want to delete (present in the catalog)
+    // remove these uuids from the Vec of all uuids we checked.
+    to_check_in_catalog
+        .iter()
+        .filter(|c| !present_in_catalog.iter().any(|existing| *existing == c.0))
```
Comment (crepererum) [inline garbage_collector/src/objectstore/checker.rs:181]: `O(n^2)`? Maybe make `present_in_catalog` a `HashSet`.

File: garbage_collector/src/objectstore/checker.rs [hunk 9]
```diff
@@ -40,96 +50,153 @@ pub(crate) async fn perform(
...
-                // We have a reference to this file; do not delete
-                debug!(
-                    location = %item.location,
-                    deleting = false,
-                    reason = "exists in catalog",
-                    "Ignoring object",
-                );
-                return Ok(false);
+) -> Vec<ObjectMeta> {
+    // to_delete is the vector we will return to the caller containing ObjectMeta we think should be deleted.
```
Comment (crepererum) [inline garbage_collector/src/objectstore/checker.rs:102]: I'm not a big fan of mutable state since it is hard to follow and a pain to refactor. It's not a blocker for this PR, but if I may suggest a slightly more Rusty way (half pseudo-code):


```rust
let (to_delete, to_check_in_catalog): (Vec<_>, Vec<_>) = items.into_iter()
  .filter(/* cutoff filter */)
  .filter(/* file suffix filter */)
  .partition(/* three-state if block starting w/ `if let Some(uuid) = file_name.unwrap()...` */);
```

File: garbage_collector/src/objectstore/checker.rs [hunk 10]
```diff
@@ -40,96 +50,153 @@ pub(crate) async fn perform(
...
-                uuid,
-                reason = "not a valid UUID",
+                reason = "not a .parquet file",
                 "Scheduling file for deletion",
             );
+            to_delete.push(candidate)
+        }
+    }
+
+    let mut present_in_catalog = Vec::with_capacity(to_check_in_catalog.len());
```
Comment (crepererum) [inline garbage_collector/src/objectstore/checker.rs:160]: Similar to my mutable state comment above, for futures:

```rust
let present_in_catalog: HashSet<_> = futures::stream::iter(to_check_in_catalog.chunks(CATALOG_BATCH_SIZE))
    .then(|...| async move {})
    .flat_map(|ids| futures::stream::iter(ids))
    .collect()
    .await;
```

File: garbage_collector/src/objectstore/checker.rs [hunk 11]
```diff
@@ -40,96 +50,153 @@ pub(crate) async fn perform(
...
+    for batch in to_check_in_catalog.chunks(CATALOG_BATCH_SIZE) {
+        let just_uuids = batch.iter().map(|id| id.0).collect();
+        match check_ids_exists_in_catalog(just_uuids, parquet_files).await {
+            Ok(mut present_uuids) => {
+                present_in_catalog.append(&mut present_uuids);
+            }
+            Err(e) => {
```
Comment (crepererum) [inline garbage_collector/src/objectstore/checker.rs:167]: I think this is wrong: if you have a catalog error (maybe IO?) then you just treat the whole batch as absent and wipe it? Like whenever you loose the connection to PG for a short while you delete the file?

**Review by crepererum — CHANGES_REQUESTED** [review]

File: data_types/src/lib.rs (hunk 5 above)
Comment (philjb) [reply data_types/src/lib.rs:686]: > need this

Which is this? The derive? A part of the derive?

File: garbage_collector/src/objectstore/checker.rs (hunk 3 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs:40]: done.

File: garbage_collector/src/objectstore/checker.rs (hunk 6 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs:44]: good point. I pulled the select up a layer where i think it wants to be anyway.

File: garbage_collector/src/objectstore/checker.rs (hunk 4 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs:74]: thanks!

File: data_types/src/lib.rs (hunk 5 above)
Comment (crepererum) [reply data_types/src/lib.rs:686]: The entire type. See my other comment: I think the affected catalog method should not contain the struct but the plain ID.

File: garbage_collector/src/objectstore/checker.rs (hunk 8 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs:181]: good call. thanks.

` .filter(|c| !present_in_catalog.contains(&c.0))`

File: garbage_collector/src/objectstore/checker.rs (hunk 11 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs:167]: yup - you're right. I'll see about adding the missing test for the negative/failure path.

File: garbage_collector/src/lib.rs [hunk 12]
```diff
@@ -129,20 +129,14 @@ impl GarbageCollector {
             }
         })?;
 
-        let checker_perform =  async || {
```
Comment (carols10cents) [inline garbage_collector/src/lib.rs]: Unfortunately async closures are unstable :(

File: garbage_collector/src/objectstore/checker.rs [hunk 13]
```diff
@@ -419,4 +433,94 @@ mod tests {
         assert_eq!(results.len(), 1);
         assert_eq!(results[0], item);
     }
+
+    struct MockParquetFileRepo {}
+
+    #[async_trait]
+    impl ParquetFileRepo for MockParquetFileRepo {
```
Comment (carols10cents) [inline garbage_collector/src/objectstore/checker.rs:440]: I wanted to make sure this actually worked, so I started a trait impl for you... but I did leave all the unused parameter warnings as an exercise for the reader ;)

File: iox_catalog/src/sqlite.rs [hunk 14]
```diff
@@ -1337,25 +1337,28 @@ WHERE object_store_id = $1;
...
-        let rec = sqlx::query_as::<_, ParquetFileExists>(
-            r#"
-SELECT 1 as exists
+    async fn exists_by_object_store_id_batch(
+        &mut self,
+        object_store_ids: Vec<Uuid>,
+    ) -> Result<Vec<ParquetFileExistsByOSID>> {
+        let in_value = object_store_ids
+            .into_iter()
+            .map(|id| format!("uuid_blob('{}')", id.hyphenated()))
```
Comment (philjb) [inline iox_catalog/src/sqlite.rs:1346]: this works for me in `sqlite3` cli but doesn't in the tests, saying the function does not exist.

```
called `Result::unwrap()` on an `Err` value: SqlxError { source: Database(SqliteError { code: 1, message: "no such function: uuid_blob" }) }
thread 'sqlite::tests::test_catalog' panicked at 'called `Result::unwrap()` on an `Err` value: SqlxError { source: Database(SqliteError { code: 1, message: "no such function: uuid_blob" }) }', iox_catalog/src/interface.rs:2167:14
stack backtrace:
```

I know that there needs to be some conversion of the string to the uuid blob type. Since I can't use binding that was doing this previously, I'm not sure how for sqlite (or sql in general) I can specify the type I want sqlite to interpret the string as. I also got `uuid_str(object_store_id)` in the where clause to work as well in the cli. I suspect that is the slower method (convert all objectstoreids to strings first).

File: iox_catalog/src/postgres.rs (hunk 7 above)
Comment (philjb) [reply iox_catalog/src/postgres.rs:1473]: > Can we just make the exists_by_object_store_id batched and not add yet another method? I think the non-batch method is only used by tests and they can easily use a 1-batch.

Yes. done.

> Also your new interface is not covered by a test (should be done in interface.rs, there are generic tests).

Yes. done. But I'm struggling to get the uuid casted/converted for sqlite.

> Also I think the method should just return a ParquetFileID or something, not a struct w/ a single element (this is just an implementation detail of the SQL wrapper).

It needs to return the object_store_id because that's what I have. Instead of the struct, just a Vec<Uuid> should be ok. I will convert.

File: garbage_collector/src/objectstore/checker.rs (hunk 9 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs:115]: That's fair. I first tried to do it with the fluet filtering and mapping but couldn't manage it and switched to what you see here. I didn't know about `partition` so perhaps I will try again after this pr to get it back into that style.

File: garbage_collector/src/objectstore/checker.rs (hunk 11 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs:181]: test added.

File: iox_catalog/src/sqlite.rs (hunk 14 above)
Comment (crepererum) [reply iox_catalog/src/sqlite.rs:1346]: We're not really using the latest sqlite:

https://github.com/influxdata/influxdb_iox/blob/614ff35998623026bd78b682dbce79827942e639/Cargo.lock#L3370-L3372

We cannot upgrade it due to sqlx (the abstraction layer that we use) has a somewhat messed up release policy.

File: garbage_collector/src/objectstore/checker.rs [hunk 15]
```diff
@@ -22,123 +24,205 @@ pub enum Error {
...
+    parquet_files: &mut dyn ParquetFileRepo,
+    cutoff: Duration,
+    mut items: mpsc::Receiver<ObjectMeta>,
+    deleter: mpsc::Sender<ObjectMeta>,
+) -> Result<()> {
+
+    let mut batch = Vec::with_capacity(CATALOG_BATCH_SIZE);
+    let mut timedout;
     loop {
-        tokio::select! {
```
Comment (philjb) [inline garbage_collector/src/objectstore/checker.rs:44]: pulled up above `checker:: perform` - see earlier comment.

File: garbage_collector/src/objectstore/checker.rs (hunk 3 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs]: added into a comment.

File: garbage_collector/src/objectstore/checker.rs (hunk 4 above)
Comment (philjb) [reply garbage_collector/src/objectstore/checker.rs]: done.

File: data_types/src/lib.rs (hunk 5 above)
Comment (philjb) [reply data_types/src/lib.rs:686]: removed the struct type, switching to an explicit map (instead of leveraging FromRow)

File: iox_catalog/src/sqlite.rs (hunk 14 above)
Comment (philjb) [reply iox_catalog/src/sqlite.rs]: addressed this with the sqlite blob literal syntax: https://www.sqlite.org/lang_expr.html#literal_values_constants_

BLOB literals are string literals containing hexadecimal data and preceded by a single "x" or "X" character. Example: X'53514C697465'

File: data_types/src/lib.rs [hunk 16]
```diff
@@ -673,14 +673,6 @@ impl From<SkippedCompaction> for compactor_proto::SkippedCompaction {
     }
 }
 
-/// Struct to contain row result from a SELECT 1 as exists FROM ... query to support
-/// testing the existence of a parquet file.
-#[derive(Debug, Copy, Clone, PartialEq, Eq, sqlx::FromRow)]
-pub struct ParquetFileExists {
```
Comment (philjb) [inline data_types/src/lib.rs:679]: This was for the non-batching sql query version, which has been removed.

File: garbage_collector/src/objectstore/checker.rs [hunk 17]
```diff
@@ -338,6 +433,139 @@ mod tests {
...
+        let results = should_delete(vec![item.clone()], cutoff, &mut mocked_parquet_files).await;
+        assert_eq!(results.len(), 0);
+    }
+
+    struct MockParquetFileRepo<'a> {
+        inner: &'a mut dyn ParquetFileRepo,
+    }
+
+    #[async_trait]
+    impl ParquetFileRepo for MockParquetFileRepo<'_> {
```
Comment (philjb) [inline garbage_collector/src/objectstore/checker.rs:494]: thanks to @carols10cents!

**Review by crepererum — APPROVED** [review]
As agreed in the comments, further clean-ups CAN be done in future PRs. Let's ship it!

**Review by domodwyer — APPROVED** [review]
Love it! Thanks for getting this in! 🙏
